### PR TITLE
#158560081 User sees fields relevant to the category they are adding an asset to

### PR DESCRIPTION
**Pull request by Dammyy**
**What does this PR do?**

User sees fields relevant to the category they are adding an asset to

**Description of Task to be completed?**

 User sees fields relevant to the category they are adding an asset to

**How should this be manually tested?**

- click on the add asset button on the dashboard page
- select a category and click continue
- then the form loads with the relevant input fields for the category

**Any background context you want to provide?**

none

**What are the relevant pivotal tracker stories?**

[#158560081](https://www.pivotaltracker.com/story/show/158560081)

**Screenshots (if appropriate)**

screen shot 2018-07-20 at 6 41 34 pm

**Questions:**

none

### CODE REVIEW

File: src/store/modules/categories/index.ts [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
+import {  GET_CATEGORIES_FAILURE, GET_CATEGORIES_SUCCESS, GET_CATEGORY_ATTRIBUTES_FAILURE,
+  GET_CATEGORY_ATTRIBUTES_SUCCESS, 
+} from './types';
+
+// helper functions
+import axiosInstance from 'utils/helpers/axiosInstance';
 
 // action creators
+import { displayToastMessage } from 'modules/toasts';
+
```
Comment (missating) [inline src/store/modules/categories/index.ts]: This should come before the helper functions `// thunks (action creators)`

File: src/store/modules/categories/index.ts [hunk 2]
```diff
@@ -39,9 +65,27 @@ export const getAssetCategories = () => (dispatch) => {
...
 
+/**
+ * Returns an object containing action type and payload
+ *
+ * @param {any} dispatch
+ * @returns {object} action type and payload
+ */
+export const getAssetCategoryAttributes = assetCategoryId => (dispatch) => {
+  // // TODO: make a call to get the data from the backend
+  return axiosInstance.get('/asset-categories/2/attributes')
```
Comment (missating) [inline src/store/modules/categories/index.ts]: Since you are already making the API call, then there should be no need for the TODO comment

File: src/utils/helpers/axiosInstance.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import axios from 'axios';
+import * as Cookie from 'cookies-js';
+
+const token = Cookie.get('jwt-token');
+
```
Comment (missating) [inline src/utils/helpers/axiosInstance.ts]: The token is already available in the `authservice` no need importing the `cookies-js` to get it here again. 

You can use the authservice to get the token

File: src/components/AddAssetForm/index.tsx [hunk 4]
```diff
@@ -5,57 +5,72 @@ import * as React from 'react';
...
-        {renderDatePicker('warranty')}
-      </div>
-      <div className="form-group">
-        <label className="form-group__label">Device Model</label>
-        <SelectBox name="model" onChange={handleInputChange} options={categories} />
+      <div>
+      {
+        props.assetAttributes.map((attribute) => {
+          {
+            if (attribute.inputControl === 'text')  {
```
Comment (PaulKariukiRimiru) [inline src/components/AddAssetForm/index.tsx]: How do you feel about using a `Switch...case` statement here?

File: src/components/AddAssetForm/index.tsx [hunk 5]
```diff
@@ -5,57 +5,72 @@ import * as React from 'react';
...
+              return (
+                <div key={attribute.id} className="form-group">
```
Comment (PaulKariukiRimiru) [inline src/components/AddAssetForm/index.tsx]: For more readable code put the `key` and `className` props in separate lines.

File: src/components/AddAssetPane/index.tsx [hunk 6]
```diff
@@ -12,22 +12,35 @@ import AddAssetForm from 'components/AddAssetForm';
 import Button from 'components/Button';
```
Comment (PaulKariukiRimiru) [inline src/components/AddAssetPane/index.tsx]: Add import comment here.

Comment (Dammyy) [reply src/components/AddAssetPane/index.tsx]: The import comment is there. Open the file to view the full file. The import comment part is cut off from view.

File: src/components/AddAssetPane/index.tsx [hunk 7]
```diff
@@ -12,22 +12,36 @@ import AddAssetForm from 'components/AddAssetForm';
...
   state = {
     category: 'Asset Category',
     showAssetForm: false,
     asset: {
-      color: 'white',
-      serialNumber: 'SDC-20239',
-      model: '',
-      warranty: moment(),
-      dateAdded: moment(),
+      customAttributes: {},
```
Comment (Billmike) [inline src/components/AddAssetPane/index.tsx]: `tag` and `location` are required as part of an asset attribute when creating an asset. I don't see them set in state here.

**Comment by andela-kakpobome** [comment]
Please resolve conflicts.

**Comment by andela-kakpobome** [comment]
Uncle @Dammyy  do you need special invites to address these conflicts sir?

**Comment by Dammyy** [comment]
@andela-kakpobome I just resolved the merge conflicts.

File: src/components/AddAssetForm/AddAssetForm.test.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  const props = {
+    handleInputChange: jest.fn(),
+    categories: categoryArray,
+    category: '',
+    asset: {
+      tag: '',
+      location: '',
+      Attributes: [],
+    },
+    assetAttributes: [
```
Comment (andela-kakpobome) [inline src/components/AddAssetForm/AddAssetForm.test.tsx]: Can't these be in a `fixtures` file?

File: src/components/AddAssetForm/AddAssetForm.test.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+          '2019',
+        ],
+      },
+    ],
+    handleCategoryChange: jest.fn(),
+  };
+
+  it('should render the AddAssetForm component', () => {
+    const wrapper = shallow(<AddAssetForm { ...props } />);
+    expect(wrapper).toMatchSnapshot();
```
Comment (andela-kakpobome) [inline src/components/AddAssetForm/AddAssetForm.test.tsx]: This is nice, but I'd still prefer to see tests for the specific fields in your additionalAttributes

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 10]
```diff
@@ -70,47 +71,107 @@ describe('AddAssetPane component', () => {
...
+          isRequired: false,
+          id: '-LFNOhCDFNtOHxyY2DsK',
+          choices: 'multiple choices',
+        },
+      ],
+    };
+
+    const wrapper = shallow(<AddAssetPane { ...props } />);
+    wrapper.setProps(newProps);
+    expect(wrapper.instance().state.asset.Attributes).toEqual({
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Why is `Attributes` spelt with a capital letter?

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 11]
```diff
@@ -70,47 +71,107 @@ describe('AddAssetPane component', () => {
...
+  it('should set Attributes state for selected asset category', () => {
+    const newProps = {
+      ...props,
+      assetCategoryAttributes: [
+        {
+          label: 'Serial number',
+          inputControl: 'text',
+          isRequired: false,
+          id: '-LFNOhCDFNtOHxyY2DsK',
+          choices: 'multiple choices',
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Shouldn't this be more representative and be an Array?

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 12]
```diff
@@ -70,47 +71,107 @@ describe('AddAssetPane component', () => {
...
       handleInputChange(inputChangeEvent);
-      expect(wrapper.state().asset[inputChangeEvent.target.name]).toBe(inputChangeEvent.target.value);
+      expect(wrapper.state().asset.tag).toBe(inputChangeEvent.target.value);
     });
   });
 
   describe('The asset-page submit button', () => {
     it('calls toggleAssetFormVisibility if the category has not been selected yet', () => {
-      const wrapper = mount(<AddAssetPane { ...props } />);
+      const wrapper = shallow(<AddAssetPane { ...props } />);
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Why this change?

File: src/components/AddAssetPane/index.tsx [hunk 13]
```diff
@@ -124,23 +144,44 @@ class AddAssetPane extends React.Component<AddAssetPaneProps> {
...
+  handleCategoryChange = (event) => {
+    const categories = this.props.allCategories;
+    const inputValue = event.target.value;
+    
+    this.setState({
+      category: inputValue,
+    });
+
+    const category = categories.find(category => category.name === inputValue);
+    this.props.getAssetCategoryAttributes(category.customAttributes);
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/index.tsx]: Why does there need to be a separate action to get this? You already have access to those custom attributes here, assuming you are using the endpoint that includes the attributes by default(if you aren't you should, so that once you get categories from state the first time, you already have access to those attributes)

File: src/components/RadioButtons/interfaces.d.ts [hunk 14]
```diff
@@ -1,11 +1,6 @@
 export interface RadioButtonsProps {
-  options: [
-    {
-      name: string;
-      value: string;
-    }
-  ];
-  name: string;
+  options: any[];
+  name?: string;
```
Comment (andela-kakpobome) [inline src/components/RadioButtons/interfaces.d.ts]: how come name is optional?

File: src/components/SelectBox/index.tsx [hunk 15]
```diff
@@ -7,13 +7,21 @@ import './SelectBox.scss';
...
-      {name && <option disabled value={name}>Choose one</option>}
-      {options.map(option => <option value={option.name} key={option.id}>{option.name}</option>)}
-    </select>
-  </div>
-));
+const SelectBox = ({ classes, options, onChange, value, name }: SelectBoxProps) => {
+  const epochTime = new Date().getTime();
+  return (
+    <div className="select-box">
+      <select name={name} onChange={onChange} value={value} className={`select-box__content ${classes}`}>
```
Comment (andela-kakpobome) [inline src/components/SelectBox/index.tsx]: Let each of these props be on separate lines Also, let the key be the `option.name`, lose the epoch time

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 16]
```diff
@@ -54,10 +55,10 @@ describe('AddAssetPane component', () => {
 
     wrapper.find('SelectBox').simulate('change', {
       target: {
-        value: 'Chairs',
+        value: 'Apple TV',
```
Comment (Orlayhemmy) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: I observed that the previous value there was 'chair' and you changed it to 'Apple TV'. using string ‘Apple TV’ is giving the impression that we can fix any string there which is not so because if it is tried, the test would fail. So its better you do something like this props.allCategories[0].name since that is where the string is coming from. It gives a better understanding.

File: src/store/modules/categories/index.ts [hunk 17]
```diff
@@ -48,6 +49,16 @@ export const getAssetCategoriesFailure = (): CategoriesActionFailure => {
   return { type: GET_CATEGORIES_FAILURE };
 };
 
+/**
+ * Get asset category attribute success action creator
+ *
+ * @returns {CategoryAttributesActionSuccess}
+ */
+export const getAssetCategoryAttributes = (Attributes: {}): CategoryAttributesActionSuccess => {
```
Comment (Orlayhemmy) [inline src/store/modules/categories/index.ts]: You are passing the custom attribute as an argument to an action and sending it to the redux state again, to what end? Since each attribute in the state already has its custom attributes, you don't need this, you can just display the custom attributes directly in the component.

File: src/components/SelectBox/index.tsx [hunk 18]
```diff
@@ -7,13 +7,21 @@ import './SelectBox.scss';
...
+        {name && <option disabled value={name}>Choose one</option>}
+        {
+          options.map((option, index) =>
+            option.name !== undefined 
+              ? <option value={option.name} key={option.id}>{option.name}</option>
```
Comment (Orlayhemmy) [inline src/components/SelectBox/index.tsx]: This line is not covered by your test. Please do well to write a test for it

**Comment by Dammyy** [comment]
@andela-kakpobome requested changes has been made.

File: src/components/SelectBox/index.tsx [hunk 19]
```diff
@@ -7,13 +7,25 @@ import './SelectBox.scss';
...
       {name && <option disabled value={name}>Choose one</option>}
-      {options.map(option => <option value={option.name} key={option.id}>{option.name}</option>)}
-    </select>
-  </div>
-));
+      {
+        options.map((option, index) =>
+          option.name !== undefined 
+            ? <option value={option.name} key={option.id}>{option.name}</option>
+            : <option value={option} key={option}>{option}</option>)
```
[the remark this answers is not in the document]
Comment (Dammyy) [reply src/components/SelectBox/index.tsx]: @Orlayhemmy are you sure it is not covered ?

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 20]
```diff
@@ -73,28 +60,29 @@ describe('AddAssetPane component', () => {
...
+    it('sets the value to state if the change is triggered by a select box or traditional input field', () => {
+      const inputChangeEvent = {
+        target: {
+          value: 'apple-tv-2015',
+          name: 'model',
+        },
+      };
 
-      expect(wrapper.state().asset[dateChangeEvent.name]).toEqual(dateChangeEvent.date);
+      handleInputChange(inputChangeEvent);
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: If you are going to do this, then you should at least check that when the select box data `changes` then the function is called. We need to know that's true.

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 21]
```diff
@@ -73,28 +60,29 @@ describe('AddAssetPane component', () => {
...
+      expect(wrapper.state().asset.attributes.model).toBe(inputChangeEvent.target.value);
     });
 
-    it('sets the value to state if the change is triggered by a select box or traditional input field', () => {
+    it('sets the value to state if tag or location input changes', () => {
       const inputChangeEvent = {
         target: {
           value: 'apple-tv-2015',
-          name: 'model',
+          name: 'tag',
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Your test case says tag and location, but you only test for tag here, also the feedback above is applicable here

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 22]
```diff
@@ -103,14 +91,45 @@ describe('AddAssetPane component', () => {
...
+      const categoryChangeEvent = {
+        target: {
+          value: props.allCategories[0].name,
+        },
+      };
 
-      wrapper.find({ submit: true }).simulate('click');
-      expect(spy).toHaveBeenCalled();
+      handleCategoryChange(categoryChangeEvent);
+      expect(wrapper.state().category).toEqual('Apple TV');
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Just to save anyone the hassle of having to glance at the fixtures file, before coming back, this shouldn't be the actual string value; and also to ensure changes to the fixture file don't make this test to fail, it instead should be, `props.allCategories[0].name`

File: src/components/SelectBox/index.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (andela-kakpobome) [reply src/components/SelectBox/index.tsx]: Although you did not specifically test for this, the . coverage is still 100%

[the remark this answers is not in the document]
Comment (Dammyy) [reply src/components/SelectBox/index.tsx]: @andela-kakpobome I actually did test for this on lines 31 - 41 in SelectBox.test.tsx. That is why we have 100% coverage

[the remark this answers is not in the document]
Comment (Orlayhemmy) [reply src/components/SelectBox/index.tsx]: Ok

File: src/components/AddAssetPane/AddAssetPane.test.tsx (hunk 21 above)
Comment (Dammyy) [reply src/components/AddAssetPane/AddAssetPane.test.tsx]: I have removed `location` because now its only tag that all asset categories have by default.
